project/BuildInfoForTestGenerator.scala

```diff
@@ -0,0 +1,35 @@
+package build
+
+import sbt._
+
+import org.scalajs.sbtplugin.Stage
+import org.scalajs.linker.ModuleKind
+
+object BuildInfoForTestGenerator {
+  /** Generate a *.scala file that contains test info as a literal string.
+   *
+   *  We need this so the test suite doesn't rely on I/O and/or resources.
+   */
+  def generate(stage: Stage, scalaVersion: String, moduleKind: ModuleKind,
+      wantSourceMaps: Boolean, sourceDir: File): Seq[File] = {
+    val trg = sourceDir / "BuildInfo.scala"
+
+    val scalaCode =
+      s"""
+      package org.scalajs.testsuite.utils
+
+      private[utils] object BuildInfo {
+        final val scalaVersion = raw\"\"\"${scalaVersion}\"\"\"
+        final val isNoModule = ${moduleKind == ModuleKind.NoModule}
+        final val isCommonJSModule = ${moduleKind == ModuleKind.CommonJSModule}
+        final val isFastOpt = ${stage == Stage.FastOpt}
+        final val isFullOpt = ${stage == Stage.FullOpt}
+        final val hasSourceMaps = $wantSourceMaps
+      }
+      """
+
+    IO.write(trg, scalaCode)
```

Comment:
All those generators should make an effort not to write to the file if it already exists and its content matches what we're about to write. Otherwise they'll be recompiled and re-optimized on every single compilation.